Overhaul gather()

crusaderky opened the pull request:
- Closes #4698
- Closes #7982
- Closes #7996
- Supersedes #5546
- DOES NOT address #7995
- DOES NOT address #7993
- Soft-blocked by #8006

crusaderky commented on distributed/client.py:
```diff
@@ -2311,22 +2311,26 @@ async def _gather_remote(self, direct, local_worker):
+            )
+            response: dict[str, Any] = {"status": "OK", "data": data}
+            if missing_keys:
+                response = await retry_operation(
+                    self.scheduler.gather, keys=missing_keys
+                )
+                if response["status"] == "OK":
+                    response["data"].update(data)
```
- #7993 
This PR does not change the previous behaviour.

crusaderky replied:
I cannot see any aesthetical refactoring in this PR that could be moved to a separate PR while preserving the functionality.

crusaderky commented on distributed/utils_comm.py:
```diff
@@ -77,36 +116,46 @@ async def gather_from_workers(
-                logger.info(
-                    "Got an unexpected error while collecting from workers: %s", e
+                r = await task
+            except (OSError, asyncio.CancelledError, asyncio.TimeoutError):
```
TODO test (blocked by #8013)

crusaderky commented on distributed/tests/test_utils_comm.py:
```diff
@@ -63,10 +61,57 @@ async def test_gather_from_workers_permissive_flaky(c, s, a, b):
+    test_worker.py::test_gather_dep_cancelled_error
+    test_worker.py::test_get_data_cancelled_error
+    https://github.com/dask/distributed/issues/8006
+    """
```
XREF #8013

fjetter replied:
How is this test related to the PR?
I'm rather conserned about this test. I don't want us to artificially cancel the tasks. The "network stack" should not be raising a CancelledError at all without a reason. Catching those can have unwanted and hard to debug consequences. This test

crusaderky commented:
This is ready for review however I'd rather have #8013 in before so that we don't muddle the waters.

hendrikmakait commented on distributed/utils_comm.py:
```diff
@@ -77,36 +116,46 @@ async def gather_from_workers(
+                    address,
                 )
-                missing_workers.add(worker)
+                for key in d[address]:
+                    missing_keys.add(key)
+                    del to_gather[key]
```
Potentially only one of these keys caused the exception. Would it make sense to attempt gathering them individually to keep the # missing keys to a minimum? 

We may also consider returning these as erring keys, not missing. However, this is out of scope.

crusaderky replied:
Changed code to mark keys as erring, meaning that gather() won't try again.
However, note that there is no transition memory->error in the scheduler; this is the same problem as with gather_dep (https://github.com/dask/distributed/issues/6705).

Yes, we *could* try a fetch the keys one by one, but it feels like a substantial complication and ultimately overkill for what, *in theory*, should be a rare-ish problem easily weeded out during development?

hendrikmakait commented on distributed/worker.py:
```diff
@@ -1307,23 +1306,31 @@ def keys(self) -> list[str]:
+            return await retry_operation(self.scheduler.who_has, keys=keys)
+
+        result, missing_keys = await gather_from_workers(
+            keys=who_has,
+            who_has=refresh_who_has,
```
As noted above, the interchangeable use of `who_has` is confusing. Renaming the parameter of `gather_from_workers` to `refresh_who_has` might help.

hendrikmakait commented on distributed/worker.py:
```diff
@@ -1707,17 +1714,11 @@ async def batched_send_connect():
+        max_connections = self.transfer_outgoing_count_limit
         # Allow same-host connections more liberally
         if (
             max_connections
```
nit: IIUC, max_connections is now always an integer, so this can be dropped. The same goes for line 1738/1739 `max_connections is not False`.

fjetter commented on distributed/utils_comm.py:
```diff
@@ -22,52 +30,83 @@
 async def gather_from_workers(
-    who_has: Mapping[str, Collection[str]],
+    keys: Iterable[str],
+    who_has: Callable[
+        [list[str]],
+        Mapping[str, Collection[str]] | Awaitable[Mapping[str, Collection[str]]],
+    ],
```
I'm not convinced this callback structure is the right approach here. I find this makes the entire mechanism even harder to reason about. Besides, this bypasses (for better or worse) missing/update_who_has mechanics on the worker.
I believe this kind of logic should be handled a layer further up the stack instead of throwing this all into this function.

crusaderky replied:
The method on the worker exclusively serves replicate, which we need to pen in to replace with AMM to begin with. The original method bypasses the whole worker missing/update_who_has system and I chose not to change that logic since it would be very complicated and ultimately throwaway.

fjetter commented on distributed/worker.py:
```diff
@@ -1307,23 +1306,31 @@ def keys(self) -> list[str]:
-            who_has=who_has, rpc=self.rpc, who=self.address
+
+        async def refresh_who_has(keys: list[str]) -> Mapping[str, Collection[str]]:
+            nonlocal first
+            if first:
+                first = False
```
why this? Isn't the point of this callback architecture to be able to update this on every iteration?

crusaderky replied:
This avoids introducing an unnecessary additional round-trip from worker to scheduler every time the user calls replicate().
If you think that code cleanliness trumps performance in this very niche use case, I'm happy to simplify the logic.

fjetter commented on distributed/tests/test_scheduler.py:
```diff
@@ -2855,56 +2855,31 @@ async def test_gather_no_workers(c, s, a, b):
+        with captured_logger("distributed.client") as client_logger:
+            assert await c.gather(x, direct=False) == 2
 
-            assert s.tasks[fin.key].who_has == {s.workers[b.address]}
-            assert a.state.executed_count == 2
-            assert b.state.executed_count >= 1
-            # ^ leave room for a future switch from `remove_worker` to `retire_workers`
+    assert sched_logger.getvalue() == "Couldn't gather keys: {'x': 'memory'}\n" * 3
+    assert "Couldn't gather 1 keys, rescheduling" in client_logger.getvalue()
```
Putting aside for a moment what changes you are proposing here. Just reading this test I believe this behavior is false.

This test is fetching data via the scheduler but is running into connection failures. However, the `Worker` is still alive (otherwise the `BatchedSend` would've been broken and the `Worker` removed). Despite of knowing that the `Worker` is alive and it merely struggles to connect, it is *rescheduling* the key? This feels like an unwarranted escalation.

fjetter reviewed (commented):
There is a lot happening in this PR. Can we break this up a bit?

crusaderky commented on distributed/utils_comm.py:
```diff
@@ -22,52 +30,83 @@
+                await asyncio.sleep(0.15)
+                busy_workers.clear()
+
+            new_who_has = who_has(list(to_gather))
```
I'll rename it to get_who_has to match Scheduler.get_who_has

fjetter commented on distributed/utils_comm.py:
```diff
@@ -77,36 +116,46 @@ async def gather_from_workers(
-                logger.info(
-                    "Got an unexpected error while collecting from workers: %s", e
+                r = await task
+            except (OSError, asyncio.CancelledError, asyncio.TimeoutError):
+                # Note: CancelledError and asyncio.TimeoutError are rare conditions
+                # that can be raised by the network stack.
+                # See https://github.com/dask/distributed/issues/8006
+                missing_workers.add(address)
```
Catching this `CancelledError` is actually slightly worse than https://github.com/dask/distributed/pull/8013
In the other PR we were talking about a well isolated task and ensured that this task could still terminate safely when it caught a CancelledError. In this case, `gather_data_from_workers` is embedded in another coroutine and a CancelledError would actually not cause this stack to unravel but would instead cause a chain of events to handle `missing_workers`.

fjetter commented on distributed/tests/test_utils_comm.py:
```diff
@@ -64,10 +63,58 @@ async def test_gather_from_workers_permissive_flaky(c, s, a, b):
+        a.block_get_data.clear()
+        b.block_get_data.clear()
+
+        fut = asyncio.create_task(
+            gather_from_workers(keys=["x"], who_has=s.get_who_has, rpc=rpc)
+        )
+        await asyncio.wait(
+            [
+                asyncio.create_task(a.in_get_data.wait()),
+                asyncio.create_task(b.in_get_data.wait()),
+            ],
+            return_when=asyncio.FIRST_COMPLETED,
+        )
+
+        tasks = {
+            task for task in asyncio.all_tasks() if "get-data-from-" in task.get_name()
+        }
+        assert tasks
+        # There should be only one task but cope with finding more just in case a
+        # previous test didn't properly clean up
+        for task in tasks:
+            task.cancel()
+
+        a.block_get_data.set()
+        b.block_get_data.set()
+        # gather_from_workers retries transparently from the other worker
+        assert await fut == ({"x": 1}, set())
```
This test actually shows that a `CancelledError` indeed does not cause the coroutines to abort but rather that something else is happening. This is actually the behavior I am concerned about.

I suggest to move this to a different PR

crusaderky commented:
As discussed online and offline:
- moved fix of #7995 to a future PR
- moved handling of CancelledError to match #8013 to a future PR
- refactored implementation without the callback
- removed cosmetic refactorings as much as possible

crusaderky commented:
@hendrikmakait @fjetter 
This is in theory complete and ready for a second round of review.
However, I'm seeing a lot of instability in CI.

These failures I've never seen before; I'll investigate:
- 1 out of 10 runs failed: test_TaskStreamPlugin (distributed.diagnostics.tests.test_task_stream)
- 1 out of 10 runs failed: test_gather_after_failed_worker (distributed.tests.test_failed_workers)
- 1 out of 10 runs failed: test_dont_steal_fast_tasks_compute_time (distributed.tests.test_steal)

These are known offenders but potentially impacted by this PR; I'll investigate:
- 1 out of 10 runs failed: test_gather_then_submit_after_failed_workers (distributed.tests.test_failed_workers)
- 1 out of 10 runs failed: test_tell_workers_when_peers_have_left (distributed.tests.test_scheduler)
- 2 out of 10 runs failed: test_chaos_rechunk (distributed.tests.test_stress)

These are known offenders which shouldn't be impacted:
- 1 out of 8 runs failed: test_closed_input_only_worker_during_transfer (distributed.shuffle.tests.test_shuffle)
- 1 out of 8 runs failed: test_closed_worker_during_transfer (distributed.shuffle.tests.test_shuffle)
- 1 out of 8 runs failed: test_crashed_worker_during_transfer (distributed.shuffle.tests.test_shuffle)
- 2 out of 8 runs failed: test_restarting_during_transfer_raises_killed_worker (distributed.shuffle.tests.test_shuffle)
- 1 out of 10 runs failed: test_file_descriptors_dont_leak[Nanny] (distributed.tests.test_client)

crusaderky commented on distributed/client.py:
```diff
@@ -2311,22 +2311,26 @@ async def _gather_remote(self, direct, local_worker):
+                    self.scheduler.gather, keys=missing_keys
+                )
+                if response["status"] == "OK":
+                    response["data"].update(data)
```
I would like to get rid of the whole functionality in #7993, so it would be throwaway work.

crusaderky commented:
I've run all tests above 10 times on each CI environment, on main and on this PR, and counted the number of failures, and I can see no regression; this is ready for final review and merge.

On a side note, this was an extremely time-consuming activity and we should urgently pen in the time to fix CI.


| test | Failures in main | Failures in this PR |
| --- | --- | --- |
| diagnostics/tests/test_task_stream.py::test_TaskStreamPlugin | 1 | 0 |
| shuffle/tests/test_shuffle.py::test_clean_after_close | 1 | 0 |
| shuffle/tests/test_shuffle.py::test_closed_input_only_worker_during_transfer | 0 | 1 |
| shuffle/tests/test_shuffle.py::test_closed_worker_during_transfer | 29 | 35 |
| shuffle/tests/test_shuffle.py::test_crashed_worker_during_transfer | 6 | 1 |
| shuffle/tests/test_shuffle.py::test_restarting_during_transfer_raises_killed_worker | 38 | 33 |
| tests/test_client.py::test_file_descriptors_dont_leak | 9 | 10 |
| tests/test_failed_workers.py::test_gather_after_failed_worker | 4 | 2 |
| tests/test_failed_workers.py::test_gather_then_submit_after_failed_workers | 9 | 1 |
| tests/test_scheduler.py::test_tell_workers_when_peers_have_left | 1 | 1 |
| tests/test_stress.py::test_chaos_rechunk | 0 | 1 |

fjetter commented on distributed/utils_comm.py:
```diff
@@ -77,36 +86,42 @@ async def gather_from_workers(
                     ),
                     operation="get_data_from_worker",
                 ),
                 name=f"get-data-from-{address}",
             )
             for address, keys in d.items()
         }
-        response: dict[str, object] = {}
-        for worker, c in coroutines.items():
+        for address, task in tasks.items():
             try:
-                r = await c
+                r = await task
```
I know this is the same as before but the task scheduling pattern here is problematic in the event of cancellation.

If `gather_from_workers` is cancelled after the tasks were created, only the task we're currently awaiting is cancelled, all others will continue running and we'll get a "never awaited foo" warning.

I _guess_ this coroutinefunction is never actually cancelled which is why we never ran into this...

The _correct_ approach to this would be to use `asyncio.gather` (or even better but not backwards copatible, the [TaskGroup of asyncio in python 3.11+](https://docs.python.org/3/library/asyncio-task.html#asyncio.TaskGroup))

I think the changes should be straight forward to something like

```python
results = asyncio.gather(tasks, return_exceptions=True)
for addr, res in results.items():
    if isinstance(res, OSError):
        ...
```

crusaderky replied:
I'll address it in a follow-up

fjetter reviewed (approved):
The above comment is not essential. I think it's good practice to cleanup our asyncio game but considering this hasn't bothered us before, I assume this thing is just never cancelled.
